src/mon/OSDMonitor.cc

```diff
@@ -3481,14 +3499,69 @@ void OSDMonitor::get_removed_snaps_range(
 
 int OSDMonitor::get_version(version_t ver, bufferlist& bl)
 {
-    if (inc_osd_cache.lookup(ver, &bl)) {
-      return 0;
-    }
-    int ret = PaxosService::get_version(ver, bl);
-    if (!ret) {
-      inc_osd_cache.add(ver, bl);
+  return get_version(ver, mon->get_quorum_con_features(), bl);
+}
+
+void OSDMonitor::reencode_incremental_map(bufferlist& bl, uint64_t features)
+{
+  OSDMap::Incremental inc;
+  bufferlist::iterator q = bl.begin();
+  inc.decode(q);
+  // always encode with subset of osdmap's canonical features
+  uint64_t f = features & inc.encode_features;
+  dout(20) << __func__ << " " << inc.epoch << " with features " << f
+	   << dendl;
+  bl.clear();
+  if (inc.fullmap.length()) {
+    // embedded full map?
+    OSDMap m;
+    m.decode(inc.fullmap);
+    inc.fullmap.clear();
+    m.encode(inc.fullmap, f | CEPH_FEATURE_RESERVED);
+  }
+  if (inc.crush.length()) {
+    // embedded crush map
+    CrushWrapper c;
+    auto p = inc.crush.begin();
+    c.decode(p);
+    inc.crush.clear();
+    c.encode(inc.crush, f);
+  }
+  inc.encode(bl, f | CEPH_FEATURE_RESERVED);
+}
+
+void OSDMonitor::reencode_full_map(bufferlist& bl, uint64_t features)
+{
+  OSDMap m;
+  bufferlist::iterator q = bl.begin();
+  m.decode(q);
+  // always encode with subset of osdmap's canonical features
+  uint64_t f = features & m.get_encoding_features();
+  dout(20) << __func__ << " " << m.get_epoch() << " with features " << f
+	   << dendl;
+  bl.clear();
+  m.encode(bl, f | CEPH_FEATURE_RESERVED);
+}
+
+int OSDMonitor::get_version(version_t ver, uint64_t features, bufferlist& bl)
+{
+  uint64_t significant_features = OSDMap::get_significant_features(features);
+  if (inc_osd_cache.lookup({ver, significant_features}, &bl)) {
+    return 0;
+  }
+  int ret = PaxosService::get_version(ver, bl);
```

Comment:
i am reviewing the call sites of `OSDMonitor::get_version(version_t ver, uint64_t features, bufferlist& bl)`. seems they don't like any return value but "0", see https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L378. if the return value is none-zero, `assert(err == 0)` and `assert(!bl.length())`, see https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L377 .

so i am confused, if the callers dislike the `-ENOENT`, why are we take all these efforts to prepare the encoded map for them ? or shall we update the callers or the callee so they are more tolerant in this case? for example, let `OSDMonitor::get_version()` spit a `dout(1)` warning message and returned reencoded bl anyway, but don't return `-ENOENT` to appease the callers.

Reply:
we wont do any reencoding effort in the case of `-NOENT` from paxos service, do we ?

`if (!ret) {`

Reply:
i think confusion comes from the reencode being inside the if (!ret) block; that's fixed now.